File: lib/services/api_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/media_item.dart';

/// Fetches media list for a device by MAC from the CMS API.
/// Uses path-style URL first (same as browser: /api/devices/{mac}/media) with MAC built into path
/// so colons are not misparsed by Uri on Android. Then falls back to query-style.
/// Response: { "data": [ { id, title, file_type, file_size, preview_url, updated_at } ] }
class ApiService {
  final String baseUrl;

  ApiService({required this.baseUrl});

  String _encodeMac(String mac) => Uri.encodeComponent(mac);

  String get _base {
    return baseUrl.endsWith('/') ? baseUrl.substring(0, baseUrl.length - 1) : baseUrl;
  }

  /// Build path-style URL from base components so MAC colons stay in path (avoids Uri.parse misparsing on Android).
  Uri _pathStyleUri(String mac) {
    final base = _base;
    final parsed = Uri.parse(base);
    final pathPrefix = parsed.path.replaceAll(RegExp(r'/$'), '');
    final path = '$pathPrefix/api/devices/$mac/media';
    return Uri(
      scheme: parsed.scheme,
      host: parsed.host,
      port: parsed.hasPort ? parsed.port : null,
      path: path,
    );
  }

  static const _timeout = Duration(seconds: 25);

  /// Try path-style first (GET /api/devices/{mac}/media), then query (GET /api/media?mac=...). Returns list or throws.
  Future<List<MediaItem>> getDeviceMedia(String mac) async {
    final base = _base;
    final pathStyle = _pathStyleUri(mac);
    final queryStyle = Uri.parse('$base/api/media').replace(queryParameters: {'mac': mac});
    final urlsToTry = [pathStyle, queryStyle];
    String? lastTriedUrl;
    ApiException? lastError;
    for (final url in urlsToTry) {
      lastTriedUrl = url.toString();
      try {
        final response = await http.get(url).timeout(_timeout);
        if (response.statusCode == 404) {
          lastError = ApiException('Not found (404). Try in browser: $lastTriedUrl', statusCode: 404);
          continue;
        }
        if (response.statusCode != 200) {
          lastError = ApiException('Server error: ${response.statusCode}', statusCode: response.statusCode);
          continue;
        }
        final body = jsonDecode(response.body) as Map<String, dynamic>;
        final data = body['data'];
        if (data == null || data is! List) return [];
        return _parseMediaList(data, base, mac);
      } catch (e) {
        lastError = ApiException(
          'Cannot reach server. Try in browser: $lastTriedUrl',
          statusCode: null,
        );
        continue;
      }
    }
    throw lastError ?? ApiException('Cannot reach server. URL: $lastTriedUrl');
  }

  List<MediaItem> _parseMediaList(List<dynamic> data, String base, String mac) {
    final list = <MediaItem>[];
    for (final item in data) {
      if (item is Map<String, dynamic>) {
        final media = MediaItem.fromJson(item);
        final serverPreview = item['preview_url'] as String?;
        final previewPath = '/media/devices/${_encodeMac(mac)}/${media.id}';
        final previewUrl = (serverPreview != null && serverPreview.isNotEmpty)
            ? serverPreview
            : '$base$previewPath';
        list.add(MediaItem(
          id: media.id,
          title: media.title,
          fileType: media.fileType,
          fileSize: media.fileSize,
          previewUrl: previewUrl,
          updatedAt: media.updatedAt,
        ));
      }
    }
    return list;
  }
}

class ApiException implements Exception {
  final String message;
  final int? statusCode;

  ApiException(this.message, {this.statusCode});

  @override
  String toString() => message;
}

File: cms_display_android/app/src/main/java/com/cms/display/api/ApiService.kt
package com.cms.display.api

import com.cms.display.model.MediaItem
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

class ApiService(private val baseUrl: String) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(25, TimeUnit.SECONDS)
        .readTimeout(25, TimeUnit.SECONDS)
        .build()

    private fun base(): String = baseUrl.removeSuffix("/")

    private fun pathStyleUri(mac: String): String {
        val base = base()
        val encodedMac = URLEncoder.encode(mac, "UTF-8")
        return "$base/api/devices/$encodedMac/media"
    }

    fun getDeviceMedia(mac: String): List<MediaItem> {
        val pathStyle = pathStyleUri(mac)
        val queryStyle = "${base()}/api/media?mac=${URLEncoder.encode(mac, "UTF-8")}"
        val urlsToTry = listOf(pathStyle, queryStyle)
        var lastError: ApiException? = null
        for (url in urlsToTry) {
            try {
                val request = Request.Builder().url(url).get().build()
                val response = client.newCall(request).execute()
                if (response.code == 404) {
                    lastError = ApiException("Not found (404). Try in browser: $url", 404)
                    response.close()
                    continue
                }
                if (response.code != 200) {
                    lastError = ApiException("Server error: ${response.code}", response.code)
                    response.close()
                    continue
                }
                val body = response.body?.string() ?: "{}"
                response.close()
                val json = JSONObject(body)
                val data = json.optJSONArray("data") ?: return emptyList()
                return parseMediaList(data, mac)
            } catch (e: Exception) {
                lastError = ApiException("Cannot reach server. Try in browser: $url", null)
            }
        }
        throw lastError ?: ApiException("Cannot reach server.", null)
    }

    private fun parseMediaList(data: org.json.JSONArray, mac: String): List<MediaItem> {
        val base = base()
        val list = mutableListOf<MediaItem>()
        for (i in 0 until data.length()) {
            val obj = data.optJSONObject(i) ?: continue
            val previewUrl = obj.optString("preview_url").takeIf { it.isNotEmpty() }
                ?: "$base/media/devices/${URLEncoder.encode(mac, "UTF-8")}/${obj.optInt("id")}"
            val scheduleSeconds: Long? = when {
                obj.has("playback_schedule_seconds") && !obj.isNull("playback_schedule_seconds") ->
                    obj.optLong("playback_schedule_seconds").takeIf { it > 0L }
                obj.has("playbackScheduleSeconds") && !obj.isNull("playbackScheduleSeconds") ->
                    obj.optLong("playbackScheduleSeconds").takeIf { it > 0L }
                else -> null
            }
            list.add(
                MediaItem(
                    id = obj.optInt("id"),
                    title = obj.optString("title").takeIf { it.isNotEmpty() },
                    fileType = obj.optString("file_type", "image/jpeg"),
                    fileSize = obj.optInt("file_size", 0).takeIf { obj.has("file_size") },
                    previewUrl = previewUrl,
                    playbackScheduleSeconds = scheduleSeconds,
                    updatedAt = obj.optString("updated_at").takeIf { it.isNotEmpty() }
                )
            )
        }
        return list
    }
}

class ApiException(message: String, val statusCode: Int?) : Exception(message)
